src/libraries/System.Net.Requests/tests/HttpWebRequestTest.cs

```diff
@@ -2077,6 +2077,103 @@ public async Task SendHttpGetRequest_WithCachePolicyBypassCache_DoNotAddHeaders(
+            await LoopbackServer.CreateClientAndServerAsync(
+                async (uri) =>
+                {
+                    HttpWebRequest request = WebRequest.CreateHttp(uri);
+                    request.Method = "POST";
+                    request.AllowWriteStreamBuffering = false;
+                    request.SendChunked = true;
+                    var stream = await request.GetRequestStreamAsync();
+                    await Assert.ThrowsAnyAsync<Exception>(() => request.GetResponseAsync());
+                },
+                async (server) => 
+                {
+                    await server.AcceptConnectionAsync(_ =>
+                    {
+                        return Task.CompletedTask;
+                    });
+                }
+            );
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task SendHttpPostRequest_WhenBufferingChanges_Success(bool buffering)
+        {
+            const string text = "Hello World!!!!\n";
+            await LoopbackServer.CreateClientAndServerAsync(
+                async (uri) =>
+                {
+                    int size = text.Length * 128 * 1024;
+                    HttpWebRequest request = WebRequest.CreateHttp(uri);
+                    request.ContentType = "application/text";
+                    if (!buffering)
+                    {
+                        request.ContentLength = size;
+                    }
+                    request.Method = "POST";
+                    request.AllowWriteStreamBuffering = buffering;
+                    using var stream = await request.GetRequestStreamAsync();
+                    for (int i = 0; i < size / text.Length; i++)
+                    {
+                        await stream.WriteAsync(new ReadOnlyMemory<byte>(Encoding.UTF8.GetBytes(text)));
+                    }
+                    await request.GetResponseAsync();
+                },
+                async (server) =>
+                {
+                    await server.AcceptConnectionAsync(async connection =>
+                    {
+                        var data = await connection.ReadRequestDataAsync();
+                        for (int i = 0; i < data.Body.Length; i += text.Length)
+                        {
+                            Assert.Equal(text, Encoding.UTF8.GetString(data.Body[i..(i + text.Length)]));
+                        }
+                        await connection.SendResponseAsync();
+                    });
+                }
+            );
+        }
+
+        [Fact]
+        public async Task SendHttpRequest_WhenNotBuffering_SendSuccess()
+        {
+            TaskCompletionSource tcs = new();
+            await LoopbackServer.CreateClientAndServerAsync(
+                async (uri) =>
+                {
+                    HttpWebRequest request = WebRequest.CreateHttp(uri);
+                    request.Method = "POST";
+                    request.ContentLength = 5;
```

Comment:
Why is content length used when non-buffering (=streaming) is on?

Reply:
Because HttpWebRequest requires it, there is no way to know how many bytes will come through the stream if the user doesn't provide it upfront.

```c#
HttpWebRequest request = WebRequest.CreateHttp("https://httpbin.org/post");
request.AllowWriteStreamBuffering = false;
request.Method = "POST";
using (var requestStream = request.GetRequestStream())
{
    byte[] data = Encoding.UTF8.GetBytes("Hello, World!");
    requestStream.Write(data, 0, data.Length);
}
using (var response = request.GetResponse())
{
    using (var responseStream = response.GetResponseStream())
    {
        byte[] buffer = new byte[1024];
        int bytesRead;
        while ((bytesRead = responseStream.Read(buffer, 0, buffer.Length)) > 0)
        {
            Console.Write(Encoding.UTF8.GetString(buffer, 0, bytesRead));
        }
    }
}
```
This code throws that exception from here on .NET Framework.
![image](https://github.com/dotnet/runtime/assets/45091041/a09d5a70-a09e-4df6-b4b9-08b8bdfc112d)